### PR TITLE
BugFix: Async race condition on agent sort

### LINKED ISSUE #840: Race condition/crash when layer contains multiple agents reading a spatial message.
Reproducible with a Release build of [precise_chemo-sensitivity2b of primage repo](https://github.com/primagesheffield/flamegpu2-neuroblastoma/tree/precise_chemo-sensitivity2b/src).

It appears as a propagated CUDA illegal memory access error.

```
CUDAEnsemble progress: 0/70terminate called after throwing an instance of 'flamegpu::exception::CUDAError'
  what():  /home/kywertheim/flamegpu2-neuroblastoma/build/_deps/flamegpu2-src/include/flamegpu/gpu/detail/CUDAErrorChecking.cuh(27): CUDA Error: /home/kywertheim/flamegpu2-neuroblastoma/build/_deps/flamegpu2-src/include/flamegpu/util/detail/CUDAEventTimer.cuh(40): an illegal memory access was encountered
Aborted (core dumped)
```

In this case, that is occurring within the call to `step()` inside the sub model.

image

So I fixed the fact a submodel within an ensemble should not have `CUDAEventTimer`, by just marking all submodels as `is_ensemble=true`.

This however just caused the propagated error to appear elsewhere.

```
rob@mavericks:~/flamegpu2-neuroblastoma/build/bin/Release$ ./fgpu2_neuroblastoma
CUDAEnsemble progress: 0/10000terminate called after throwing an instance of 'flamegpu::exception::CUDAError'
terminate called recursively
terminate called recursively
  what():  /home/rob/flamegpu2-neuroblastoma/build/_deps/flamegpu2-src/include/flamegpu/gpu/detail/CUDAErrorChecking.cuh(27): CUDA Error: /home/rob/flamegpu2-neuroblastoma/build/_deps/flamegpu2-src/src/flamegpu/exception/FLAMEGPUDeviceException.cu(17): an illegal memory access was encountered
Aborted (core dumped)
````

So I updated `CUDASimulation::synchronizeAllStreams()` to recursivley synchronize streams in submodels too and additionally called it before calling `simulate()` on a submodel. However this made no difference.

So it's still a mystery requiring more investigation.

Current backtrace implies it's todo with destruction of `CUDASimulation` of a submodel.

```
terminate called after throwing an instance of 'flamegpu::exception::CUDAError'
terminate called recursively
  what():  /home/rob/flamegpu2-neuroblastoma/build/_deps/flamegpu2-src/include/flamegpu/gpu/detail/CUDAErrorChecking.cuh(27): CUDA Error: /home/rob/flamegpu2-neuroblastoma/build/_deps/flamegpu2-src/src/flamegpu/exception/FLAMEGPUDeviceException.cu(17): an illegal memory access was encountered
terminate called recursively

Thread 13 "CUDASim D2T0" received signal SIGABRT, Aborted.
[Switching to Thread 0x7fffb57fe000 (LWP 2262821)]
__GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
50      ../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) backtrace
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007ffff3de1859 in __GI_abort () at abort.c:79
#2  0x00007ffff41ec53a in __gnu_cxx::__verbose_terminate_handler() () from /lib/x86_64-linux-gnu/libstdc++.so.6
#3  0x00007ffff41ea38c in ?? () from /lib/x86_64-linux-gnu/libstdc++.so.6
#4  0x00007ffff41e9369 in ?? () from /lib/x86_64-linux-gnu/libstdc++.so.6
#5  0x00007ffff41e9d21 in __gxx_personality_v0 () from /lib/x86_64-linux-gnu/libstdc++.so.6
#6  0x00007ffff3fe6bef in ?? () from /lib/x86_64-linux-gnu/libgcc_s.so.1
#7  0x00007ffff3fe7281 in _Unwind_RaiseException () from /lib/x86_64-linux-gnu/libgcc_s.so.1
#8  0x00007ffff41ea69c in __cxa_throw () from /lib/x86_64-linux-gnu/libstdc++.so.6
#9  0x00005555555b5b6b in flamegpu::exception::DeviceExceptionManager::~DeviceExceptionManager() ()
#10 0x000055555561fddf in flamegpu::CUDASimulation::~CUDASimulation() ()
#11 0x0000555555630dec in std::_Rb_tree<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::unique_ptr<flamegpu::CUDASimulation, std::default_delete<flamegpu::CUDASimulation> > >, std::_Select1st<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::unique_ptr<flamegpu::CUDASimulation, std::default_delete<flamegpu::CUDASimulation> > > >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::unique_ptr<flamegpu::CUDASimulation, std::default_delete<flamegpu::CUDASimulation> > > > >::_M_erase(std::_Rb_tree_node<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::unique_ptr<flamegpu::CUDASimulation, std::default_delete<flamegpu::CUDASimulation> > > >*) ()
#12 0x000055555561f61f in flamegpu::CUDASimulation::~CUDASimulation() ()
#13 0x000055555561fed9 in flamegpu::CUDASimulation::~CUDASimulation() ()
#14 0x000055555565c922 in flamegpu::SimRunner::start() ()
#15 0x00007ffff4216de4 in ?? () from /lib/x86_64-linux-gnu/libstdc++.so.6
#16 0x00007ffff3fb9609 in start_thread (arg=<optimised out>) at pthread_create.c:477
#17 0x00007ffff3ede163 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
(gdb)

```

WIP hacking to fix it here: https://github.com/FLAMEGPU/FLAMEGPU2/tree/async_submodel_crash

### PR DESCRIPTION
**Pull request by Robadob**
Various other important improvements. I'm happy for the commits to remain separate as they should all be distinct changes.

Still need to actually do the big bugfix, but wanted to bookmark the branch,

Closes #840

### CODE REVIEW

**Comment by Robadob** [comment]
Windows, CUDA 11.4, Release, Seatbelts=ON

```
[----------] Global test environment tear-down
[==========] 1038 tests from 81 test suites ran. (572003 ms total)
[  PASSED  ] 1038 tests.

  YOU HAVE 5 DISABLED TESTS
```

Still need to test it does fix the primage crash. But think this should tie up the known async issues.

File: include/flamegpu/gpu/CubTemporaryMemory.cuh [hunk 1]
```diff
@@ -0,0 +1,24 @@
+#ifndef INCLUDE_FLAMEGPU_GPU_CUBTEMPORARYMEMORY_CUH_
```
Comment (ptheywood) [inline include/flamegpu/gpu/CubTemporaryMemory.cuh:1]: `CubTemporaryMemory` should be in `detail`, it is not part of the user-interface (this is also true of things like CUDAScanCompaction, but that isn't being added by this PR). 

Alternatively the header could be placed in the `src` directory, but that would depend on where it is included from (so would probably require more substantial changes)

**Review by ptheywood — CHANGES_REQUESTED** [review]

File: include/flamegpu/gpu/CubTemporaryMemory.cuh (hunk 1 above)
Comment (Robadob) [reply include/flamegpu/gpu/CubTemporaryMemory.cuh:1]: Most of the `gpu` sub dir is the same. Have addressed this.

File: src/flamegpu/gpu/CUDASimulation.cu [hunk 2]
```diff
@@ -2074,10 +2078,13 @@ void CUDASimulation::destroyStreams() {
 }
 
 void CUDASimulation::synchronizeAllStreams() {
-    // Destroy streams.
+    // Sync streams.
     for (auto stream : streams) {
         gpuErrchk(cudaStreamSynchronize(stream));
     }
+    for (auto &it : submodel_map) {
```
[the remark this answers is not in the document]
Comment (Robadob) [reply src/flamegpu/gpu/CUDASimulation.cu]: Fair, they are getting sync'd end of every layer. All that occurs after that is host api stuff (exit fns/logging etc). This is just leftover from my debugging, have removed.

**Review by ptheywood — APPROVED** [review]

**Pull request merged**